third_party/amd/lib/TritonAMDGPUTransforms/StreamPipeline.cpp

```diff
@@ -196,54 +205,70 @@ class StreamPipeliner {
-  } else if (isMultibuf) {
-    // Streaming Schema cluster order and staging for multi-buffer.
-    // for i in (...):
-    //   local_stores: stage=i+1
-    //   global_loads: stage=i+2
-    //   local_load:   stage=i
-    //   compute:      stage=i
-    //   tail:         stage=i
-    config[SCHED_LOCAL_STORE] = {lastStage - 1, schedule.clusters.newAtBack()};
-    auto cluster1 = schedule.clusters.newAtBack();
-    config[SCHED_GLOBAL_LOAD] = {0, cluster1};
-    config[SCHED_LOCAL_LOAD] = {lastStage, cluster1};
-    config[SCHED_COMPUTE] = {lastStage, cluster1};
-    config[SCHED_TAIL] = {lastStage, schedule.clusters.newAtBack()};
-  } else {
-    // Streaming Schema cluster order and staging for single-buffer.
-    // for i in (...):
-    //   global_loads: stage=i+1
-    //   local_load:   stage=i
-    //   compute:      stage=i
-    //   local_stores: stage=i+1
-    //   tail:         stage=i
-    auto cluster0 = schedule.clusters.newAtBack();
-    config[SCHED_GLOBAL_LOAD] = {0, cluster0};
-    config[SCHED_LOCAL_LOAD] = {lastStage, schedule.clusters.newAtBack()};
-    config[SCHED_COMPUTE] = {lastStage, cluster0};
-    config[SCHED_LOCAL_STORE] = {lastStage - 1, schedule.clusters.newAtBack()};
-    config[SCHED_TAIL] = {lastStage, schedule.clusters.newAtBack()};
+LogicalResult StreamPipeliner::initSchedule(int maxIndirectionLevel) {
+
+  bool pairedGlobalLoadLocalStore = stages[SCHED_LOCAL_STORE] == 0;
+  stages[SCHED_LOCAL_STORE] += maxIndirectionLevel;
+
+  if (stages[SCHED_LOCAL_STORE] >= numStages ||
+      stages[SCHED_LOCAL_STORE] > stages[SCHED_LOCAL_LOAD]) {
+    LDBG("Invalid schedule: GLOBAL_LOAD stage = "
+         << stages[SCHED_GLOBAL_LOAD]
+         << ", LOCAL_STORE stage = " << stages[SCHED_LOCAL_STORE]
+         << ", LOCAL_LOAD stage = " << stages[SCHED_LOCAL_LOAD]
+         << ", COMPUTE stage = " << stages[SCHED_COMPUTE]);
+    return failure();
+  }
+
+  // Calculate the number of buffers needed for each load.
+  // TODO: Use the precise number of buffers needed by the particular load.
+  numBuffers =
+      std::max(1, stages[SCHED_LOCAL_LOAD] - stages[SCHED_LOCAL_STORE]);
+  LDBG("deduced max shared memory buffer number = " << numBuffers);
+
+  // If tt.load and ttg.local_store are in the same stage
+  //   spread them apart to allow overlap with compute
+  // else
+  //   Initiate ttg.local_store before tt.load
+  int globalLoadCluster = 0;
+  int localStoreCluster = 2;
+  if (!pairedGlobalLoadLocalStore) {
+    globalLoadCluster = 2;
+    localStoreCluster = 1;
+  }
+
+  // If ttg.local_load and ttg.local_store are in the same stage
+  //   spread them apart to allow overlap with compute
+  // else if they share the buffer
+  //   ttg.local_load must come first
+  // else
+  //   schedule ttg.local_load in the middle
+  int localLoadCluster = globalLoadCluster;
+  if (stages[SCHED_LOCAL_LOAD] == stages[SCHED_LOCAL_STORE]) {
+    localLoadCluster = std::max(2, localStoreCluster + 1);
```

Comment:
Is the `max(2` part useful? From the above `localStoreCluster + 1` is guaranteed to be >= 2.